### ISSUE TITLE
Support URL form encoded request bodies

### ISSUE THREAD

**Issue by benfrearson** [apple/swift-openapi-generator#182]
There may be a legitimate reason why this isn't already implemented, or perhaps I'm not using the generator for its intended use case - if so, I'd appreciate some clarity.

It appears to me that URL form encoded requests have their data structure erased in the generated Input.body type, which blindly expects `Data`.

``` 
# openapi.yaml
paths:
  /login:
    post:
      tags:
        - Auth
      summary: Login
      description: Login
      operationId: login
      requestBody:
        required: true
        content:
          application/x-www-form-urlencoded:
            schema:
              type: object
              required:
                - device_name
                - email
                - password
              properties:
                device_name:
                  type: string
                email:
                  type: string
                password:
                  type: string
```

```
//Types.swift

//...

/// API operations, with input and output types, generated from `#/paths` in the OpenAPI document.
public enum Operations {
    /// Login
    ///
    /// Login
    ///
    /// - Remark: HTTP `POST /login`.
    /// - Remark: Generated from `#/paths//login/post(login)`.
    public enum login {
        public static let id: String = "login"
        public struct Input: Sendable, Equatable, Hashable {
            
           // ...

            @frozen public enum Body: Sendable, Equatable, Hashable { case binary(Foundation.Data) }
            public var body: Operations.login.Input.Body
            /// Creates a new `Input`.
            ///
            /// - Parameters:
            ///   - path:
            ///   - query:
            ///   - headers:
            ///   - cookies:
            ///   - body:
            public init(
                path: Operations.login.Input.Path = .init(),
                query: Operations.login.Input.Query = .init(),
                headers: Operations.login.Input.Headers = .init(),
                cookies: Operations.login.Input.Cookies = .init(),
                body: Operations.login.Input.Body
            ) {
                self.path = path
                self.query = query
                self.headers = headers
                self.cookies = cookies
                self.body = body
            }
        }
// ...
}
```

The issue here is that using the corresponding `Client` method now is unaware of the required url form fields. This is obviously fairly easy to work around, but it obfuscates the expected data from the users of the `Client` in a way that using a JSON encoded body does not.

It would be better if the generated Input type could handle encoding form bodies itself, such as by:
- Having an `Input.body` initializer that accepts arguments as dictated by the form fields.
- Having an intermediate property for the key-value pairs that is encoded when serializing the request similar to the `converter.setRequiredRequestBodyAsBinary()` method.

I would be happy to contribute to this issue, but I'd like to confirm that this is something that is needed/wanted and aligns with the project's direction.

**Comment by czechboy0** [comment]
Hi @benfrearson, welcome to the project! Thanks for taking the time to file the issue with so much great detail!

You're right, Swift OpenAPI Generator doesn't yet support URL encoded form bodies as a structured type, but it's definitely something we'd accept a PR for! It just hasn't been implemented yet as we're still working through other foundational OpenAPI features.

Since adding support for a new kind of a structured body will not be a trivial amount of work, the best way might be for you to try to prototype it (even including some shortcuts) and open a draft PR that we can then discuss the approach in. I don't want you to spend time polishing a PR until we confirm that the overall approach is likely to be accepted.

If this sounds good to you, I'll be looking forward to your PR and if you have any questions, you can ask under this issue.

Thanks! 🙏

**Comment by czechboy0** [comment]
Also, this piece of documentation could be helpful, you'll potentially need to expand the set of helper functions: https://swiftpackageindex.com/apple/swift-openapi-generator/0.1.8/documentation/swift-openapi-generator/converting-between-data-and-swift-types

**Comment by benfrearson** [comment]
Thanks! Yes, I was anticipating using the Converter to handle this. How does this sound as a general approach from the Client interface side:

- Add an additional initializer to the Input.body type that's derived from the defined form entries in the openapi document.
    - eg `init (_ formBody: Components.Schemas.LoginRequest)`
    - For a schema defined directly within the response object and not using `#ref` we would embed a new `URLFormRequest` struct in the `Input` type.

### Assumptions:
- The majority of URL form encoding uses a flat key-value pair structure
- Provided schemas may not cover all possible keys, and must allow for arbitrary values
- Some endpoints may not follow the convention and encode data in an unexpected way. E.g.:
    - nested data types in schema (possibly expecting stringified data)
    - repeated key values (e.g. `tag=article&tag=sports`)

### Considerations/scope:
- Must provide an interface of known (required and optional) form fields to `Input.Body`
- Must allow for arbitrary fields not included in the schema
- Must not break current implementation (keep the current implicit initializer: `.binary(Data)`)
- Should consider unconventional uses of form data (though arguably these are covered by the current implementation)
    - nest data types might have an option to stringify or use dot-syntax to encode
    
I think for a first-pass at this, we should focus on conventional url form key-value pairs while maintaining the existing implementation. I assume this would cover a large percentage of endpoints that use url forms. We could extend this in the future with helpers or options for other encoding types.

**Comment by benfrearson** [comment]
On the queries about arbitrary field names: I thought it feasible—but probably very bad practice—that you could design an endpoint which would accept any field name, regardless of whether it's included in a schema. 

For example, you might have a `api/customAttributes` endpoint where a user could submit any field names they wanted: 

```
customFieldName1=true&customFieldName2=false`
```

Again, I don't have any concrete examples of this, and it would be fairly odd imo, but I think it would be possible to write an endpoint that did this (however weird)...

---
We can absolutely aim for parity with JSON support, and that would be preferable in the particular instance that drew me here: I agree that we want as much type safety as possible. 

My only hesitation with this approach is that we're making the assumption that every form request is equally as uniform as a JSON object. 

I'm happy to move ahead with this approach though and see if it causes any actual issues. The edge cases I suggested are feasible, but perhaps minimal, and it may be good to get the ball rolling with covering the best practices first.

**Comment by czechboy0** [comment]
Thinking about this more, creating a proper implementation of a URL form encoder and decoder (for `Codable` types) could have pretty substantial positive consequences for this project, as (with slightly different flags enabled/disabled), it could be used both for `application/x-www-form-urlencoded` bodies, and for encoding/decoding complex query items and headers themselves, and it's scale to the full support of JSON Schema-representable types, not just individual types we support today (through the `_StringConvertible` protocol and a bunch of manually written overloads in the runtime library).

To that end, I'd suggest we focus on that problem first - can we build a `URLFormEncoder` and `URLFormDecoder` types, which have a few options that allow us to use them both in query items and in bodies (such as whether arrays use indexes or not: `key=value1&key=value2` vs `key[0]=value1&key[1]=value2`, whether spaces are encoded as `+` or `%20`, and possibly a few more)?

Once we have a decent encoder and decoder in the runtime library, including unit tests, we can discuss how to a) add support for the new content types in the generator itself (should be straightforward at that point) and b) how to migrate the existing query item and header encoding logic over to this new infra (can be done separately, but should result in a bunch of code removed).

This might take a few PRs, so we could introduce it in the runtime library under `@_spi(URLFormCoding)` while we work on it, to make sure no adopters start relying on it prematurely, and build it up from there.

I took a brief look at Foundation's JSONEncoder/JSONDecoder [implementation](https://github.com/apple/swift-foundation/blob/main/Sources/FoundationEssentials/JSON/JSONEncoder.swift) and I think they have a good structure we could get inspired by, with the difference that the URLFormEncoder/Decoder should be even simpler, as we really only have 4 fundamental types (string, array, dictionary, and null), whereas JSON has a few more.

How does this approach sound?

**Comment by czechboy0** [comment]
Splitting the creation of the new type into a separate issue, to split it from actually supporting the new content type in bodies: https://github.com/apple/swift-openapi-generator/issues/192

**Comment by bfrearson** [comment]
I've just put up a first pass attempt at implementing a rudimentary version with hardcoded converters, but I fully agree that having a custom encoder is preferable and something that we can work on next. I'm happy to continue with that.

I have a few more thoughts, but will add these to #192

**Comment by czechboy0** [comment]
An update from my side: the URI coder will be landing in https://github.com/apple/swift-openapi-runtime/pull/41. Once that lands, I'll move all of the parameters to it (already WIP), deprecate a bunch of code, and we'll release a breaking 0.2.0 hopefully some time next week. After that's released, all the pieces should be in place for you to pick this work up again, and add URL form encoded body support. I'll let you know when it's ready, I'd like to avoid creating a bunch of merge conflicts in on your branches, so it's better to wait.

**Comment by bfrearson** [comment]
Sweet, thanks! I've been away for the past week, but I'll take a look at this later in the week.

**Comment by czechboy0** [comment]
Landed behind the feature flag `urlEncodedForm`, soon to be released in the next 0.2.x release.
